src/node.rs

```diff
@@ -0,0 +1,219 @@
+use crate::helper::Identifier;
+use crate::kbucket::KbucketTable;
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, SocketAddrV4};
+use tokio::io;
+use tokio::net::UdpSocket;
+
+#[derive(Debug)]
+pub enum FindNodeResult {
+    Found(Option<TableRecord>),
+    NotFound(Vec<Option<TableRecord>>),
+}
+
+#[derive(Debug)]
+pub enum Search {
+    Success(usize, usize),
+    Failure(usize, usize),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct TableRecord {
+    pub node_id: Identifier,
+    pub ip_address: Ipv4Addr,
+    pub udp_port: u16,
+    pub socket_addr: SocketAddrV4,
+}
+
+// The main Kademlia client struct.
+// Provides user-level API for performing querie and interacting with the underlying service.
+// #[derive(Clone, Copy, Debug, PartialEq)]
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    pub node_id: Identifier,
+    pub table_record: TableRecord,
+    pub table: KbucketTable,
```

Comment:
why does a `TableRecord` live outside the `KBucketTable`?

to first order, the table should just be a map from `node_id`s to pairs `(ip, port)`, with some extra logic to batch them by `xor` distance

I'd try to simplify as much as possible

Reply:
you can structure however you want, but the name `TableRecord` implies it is a thing inside the table

all a node really needs to know about itself is its `node_id` I think -- do you send over your (ip, port) in a way that you can't just learn once at startup and cache in e.g. an ENR or something?

Reply:
> 

I don't think so

Reply:
> why does a `TableRecord` live outside the `KBucketTable`?
> 
> to first order, the table should just be a map from `node_id`s to pairs `(ip, port)`, with some extra logic to batch them by `xor` distance
> 
> I'd try to simplify as much as possible

Is this design simpler bc std::collections::Hashmap exposes certain functionalities for me?  ie store and get